CODING_RULES.md

```diff
@@ -11,90 +11,156 @@ to be completed
+
+- on the node: in general, node should not stop because of exceptions that occur while executing requests received from researcher
+
+- on the researcher: general behaviour is to propagate the exceptions to the top level layer, where they are transformed to a friendlier output. Researcher displays this output and stops.
+
+- when a class raises an exception, it raises a FedbiomedSomethingError: use exceptions defined in **fedbiomed.common.exceptions** :
+
+  Do:
+  ```
+  raise FedbiomedSometypeError()
+  ```
+
+  Don't:
+  ```
+  raise NameError()
+  ```
+
+- optionally, if more specificity is wanted, a class can catch a python (non-Fed-Biomed) exception and re-raise it as a FedbiomedError
+
+  Optionally do:
+  ```
+  try:
+      something()
+  except SysError as e:
+      raise OneOfFedbiomedError()
+  ```
+
+- a class generally shouldn't catch a FedbiomedError and re-raise (a FedbiomedError):
+
+  Don't:
+  ```
+  try:
+    somecode()
+  except FedbiomedSometypeError;
+    raise FedbiomedOnetypeError()
+  ```
+
+- when catching exceptions
+
+  - try to be specific about the exception if easy/meaningful:
+
+    Ideally:
+    ```
+    try:
+      mycode()
+    except ValueError as e:
+      ...
+    ```
+
+    If needed:
+    ```
+    try:
+      mycode()
+    except Exception as e:
+      ...
+    ```
+
+  - don't use the `except:` clause
+
+    Don't
+    ```
+    try:
+      mycode()
+    except:
+      ...
+    ```
+
+  - should separate FedbiomedError and other exceptions
+
+    Do:
+    ```
+    try:
+      mycode()
+    except FedbiomedSomeError as e:
+      ...
+    except FedbiomedError as e:
+      ...
+    except Exception as e:
+      ...
+    ```
+
+    Don't (in case where the Exception can be a FedbiomedError)
+    ```
+    try:
+      some_code()
+    except Exception as e:
+      ...
+    ```
+
+- in general, a class shouldn't log in logger when raising or re-raising an exception. The class should log when catching and not re-raising a FedbiomedError exception. The class can log when catching a and not re-raising a python exception.
+
+  Do:
+  ```
+  try:
+    some_function()
+  except FedbiomedSomeError:
+    logger.xxx(message)
+  ```
+
+  Can do:
+  ```
+  try:
+    some_function()
+  except OSError:
+    logger.xxx(message)
+  ```
+
+  Don't:
+  ```
+  try:
+    some_function()
+  except AnException:
+    logger.xxx(message)
+    raise FedbiomedSomeException()
+  ```
+
+- keep the **try:** block is as small as possible
 
 - string associated to the exception:
 
   - comes from the fedbiomed.common.constants.ErrorNumbers
 
-  - complemented by a usefull (more precise) information:
+  - can be complemented by a useful (more precise) information:
 
   => consider ErrorNumbers as categories
 
-```
-try:
-    something()
-
-except SomeError as e:
-
-    _msg = ErrorNumbers.FBxxx.value + ": the file " + filename + " does not exist"
-
-    logger.error(_msg)
-    raise OneOfFedbiomedError(_msg)
-```
+  ```
+  try:
```

Comment:
we could specify the import

`from fediomed.common.constants import ErrorNumbers`

Reply:
Added imports for error numbers